include/boost/utility/string_ref.hpp

```diff
@@ -424,48 +425,66 @@ namespace boost {
-                os.write(fill_chars, static_cast< std::size_t >(chunk_size));
-            if (n > 0 && os.good())
-                os.write(fill_chars, n);
+        inline bool sr_os_fill(std::basic_ostream<charT, traits>& os, std::streamsize size)
+        {
+            enum { chunk = 8 };
+            charT fill[chunk];
+            std::fill_n(fill, static_cast<std::size_t>(chunk), os.fill());
+            for (; size >= chunk; size -= chunk) {
+                if (os.rdbuf()->sputn(fill, chunk) != chunk) {
+                    return false;
+                }
             }
+            return !(size > 0) || os.rdbuf()->sputn(fill, size) == size;
+        }
 
         template<class charT, class traits>
-        void sr_insert_aligned(std::basic_ostream<charT, traits>& os, const basic_string_ref<charT,traits>& str) {
-            const std::size_t size = str.size();
-            const std::size_t alignment_size = static_cast< std::size_t >(os.width()) - size;
-            const bool align_left = (os.flags() & std::basic_ostream<charT, traits>::adjustfield) == std::basic_ostream<charT, traits>::left;
-            if (!align_left) {
-                detail::sr_insert_fill_chars(os, alignment_size);
-                if (os.good())
-                    os.write(str.data(), size);
-                }
-            else {
-                os.write(str.data(), size);
-                if (os.good())
-                    detail::sr_insert_fill_chars(os, alignment_size);
+        class sr_os_holder {
+        public:
+            explicit sr_os_holder(std::basic_ostream<charT, traits>& os)
+                : os_(&os) { }
+            ~sr_os_holder() BOOST_NOEXCEPT_IF(false) {
+                if (os_) {
+                    os_->setstate(std::basic_ostream<charT, traits>::badbit);
                 }
             }
+            void release() {
+                os_ = 0;
+            }
+        private:
+            sr_os_holder(const sr_os_holder&);
+            sr_os_holder& operator=(const sr_os_holder&);
+            std::basic_ostream<charT, traits>* os_;
+        };
 
         } // namespace detail
 
     // Inserter
     template<class charT, class traits>
     inline std::basic_ostream<charT, traits>&
     operator<<(std::basic_ostream<charT, traits>& os, const basic_string_ref<charT,traits>& str) {
-        if (os.good()) {
-            const std::size_t size = str.size();
-            const std::size_t w = static_cast< std::size_t >(os.width());
-            if (w <= size)
-                os.write(str.data(), size);
-            else
-                detail::sr_insert_aligned(os, str);
-            os.width(0);
+        typedef std::basic_ostream<charT, traits> stream;
+        detail::sr_os_holder<charT, traits> hold(os);
+        typename stream::sentry entry(os);
+        if (entry) {
+            std::streamsize width = os.width();
+            std::streamsize size = str.size();
```

Comment:
`str.size()` is `std::size_t`, this line needed not to change. IMHO, the original code was better in that it explicitly casted `std::streamsize` to `std::size_t` and then proceeded with unsigned integers everywhere since width should not be negative anyway.

Reply:
Updated. Using `std::size_t` also avoids requiring `ios` in favor of `iosfwd`.